File: apps/ios/MagicMobile/OnDeviceDiagnostics.swift
import Foundation
import Combine

/// Latest local engine failure only, plus the app's own crash and hang summaries from MetricKit.
/// No snapshots, action payloads, networking, or automatic sharing.
@MainActor
final class OnDeviceDiagnostics: ObservableObject {
    static let maxBytes = 65_536
    /// What the report sheet shows and shares: the engine failure, then any crash and hang summaries.
    @Published private(set) var report: String?
    private var engineReport: String?
    private var persistedReport: String?
    private var systemReport: String?
    @Published private(set) var isHistorical = false
    @Published var errorMessage: String?
    private let directory: URL
    private var captureGeneration = 0
    private var clearing = false
    private var systemReportsObserver: AnyCancellable?
    private var file: URL { directory.appendingPathComponent("latest-engine-failure.txt") }

    nonisolated static var defaultDirectory: URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("MagicMobile/Diagnostics", isDirectory: true)
    }

    init(directory: URL? = nil) {
        self.directory = directory ?? Self.defaultDirectory
        do {
            if FileManager.default.fileExists(atPath: file.path) {
                let size = try file.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? 0
                guard size <= Self.maxBytes else { throw CocoaError(.fileReadTooLarge) }
                engineReport = try String(contentsOf: file, encoding: .utf8)
                persistedReport = engineReport
                isHistorical = true
            }
        } catch { errorMessage = "Could not load the saved engine report: \(error.localizedDescription)" }
        systemReport = OnDeviceSystemReports.section(OnDeviceSystemReports.load(directory: self.directory))
        publish()
        // MetricKit delivers payloads shortly after launch, after this store exists.
        systemReportsObserver = NotificationCenter.default.publisher(for: OnDeviceSystemReports.didChange)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] note in
                MainActor.assumeIsolated {
                    guard let self, (note.object as? URL)?.standardizedFileURL == self.directory.standardizedFileURL else { return }
                    self.systemReport = OnDeviceSystemReports.section(OnDeviceSystemReports.load(directory: self.directory))
                    self.publish()
                }
            }
    }

    private func publish() {
        let parts = [engineReport, systemReport].compactMap { $0 }
        report = parts.isEmpty ? nil : parts.joined(separator: "\n\n")
        // Crash and hang summaries always describe an earlier session.
        if engineReport == nil { isHistorical = systemReport != nil }
    }

    func beginAttempt() {
        captureGeneration += 1
        if report != nil { isHistorical = true }
    }

    func save(engineReport: String, status: String) throws {
        // Reopening the sheet must not date an existing incident as a new failure.
        if let report = self.engineReport, persistedReport == report, let range = report.range(of: "\n\n"),
           engineReport.hasPrefix(String(report[range.upperBound...])) { return }
        let version = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "unknown"
        let build = Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "unknown"
        let text = """
        MagicMobile local engine diagnostic
        App: \(version) (\(build))
        OS: \(ProcessInfo.processInfo.operatingSystemVersionString)
        Captured: \(ISO8601DateFormatter().string(from: Date()))
        Status: \(status.prefix(200))
        Private: error text may contain card information. Share only by explicit choice.

        \(String(decoding: engineReport.utf8.prefix(Self.maxBytes), as: UTF8.self))
        """
        var bytes = Data(text.utf8.prefix(Self.maxBytes))
        // A byte limit may land inside a Unicode scalar; trim only that incomplete tail.
        while String(data: bytes, encoding: .utf8) == nil { bytes.removeLast() }
        self.engineReport = String(data: bytes, encoding: .utf8)
        isHistorical = false
        errorMessage = nil
        publish()
        do {
            try Self.prepare(directory)
            #if os(iOS)
            try bytes.write(to: file, options: [.atomic, .completeFileProtection])
            #else
            try bytes.write(to: file, options: .atomic)
            #endif
            persistedReport = self.engineReport
        } catch {
            errorMessage = "Report is available to share now, but could not be saved on this device: \(error.localizedDescription)"
            throw error
        }
    }

    /// Creates the private, backup-excluded report directory.
    nonisolated static func prepare(_ directory: URL) throws {
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true,
                                                attributes: [.posixPermissions: 0o700])
        var excludedDirectory = directory
        var values = URLResourceValues(); values.isExcludedFromBackup = true
        try excludedDirectory.setResourceValues(values)
    }

    /// Deletes the engine report and the crash and hang summaries.
    func clear() throws {
        captureGeneration += 1
        if FileManager.default.fileExists(atPath: file.path) { try FileManager.default.removeItem(at: file) }
        engineReport = nil; persistedReport = nil
        try OnDeviceSystemReports.clear(directory: directory)
        systemReport = nil
        errorMessage = nil; isHistorical = false
        publish()
    }

    func capture(status: String, read: () async throws -> String?) async {
        guard !clearing else { return }
        let generation = captureGeneration
        do {
            guard let report = try await read(), generation == captureGeneration else { return }
            try save(engineReport: report, status: status)
        } catch {
            guard generation == captureGeneration else { return }
            if errorMessage == nil { errorMessage = "Could not capture the local engine report: \(error.localizedDescription)" }
        }
    }

    func clear(engine: () async throws -> Void) async throws {
        guard !clearing else { return }
        captureGeneration += 1; clearing = true
        defer { clearing = false }
        try await engine()
        try clear()
    }
}

/// The app's own crash and hang reports, as MetricKit delivers them to this phone (iOS), kept
/// only in the diagnostics directory and shared only through the report sheet. A summary holds
/// the MetricKit time window, app and OS versions, CPU architecture, the exception or hang
/// duration, and the affected thread's frames (binary name, offset and UUID). MetricKit's
/// region format, device model and process ID are left out.
enum OnDeviceSystemReports {
    struct Entry: Codable, Equatable {
        let kind: String
        let text: String
    }

    static let didChange = Notification.Name("MagicMobile.OnDeviceSystemReports.didChange")
    static let fileName = "crash-and-hang-reports.json"
    static let limit = 5
    static let maxEntryCharacters = 8_000
    static let maxFrames = 32

    /// Crash and hang summaries in one `MXDiagnosticPayload.jsonRepresentation()`.
    static func entries(fromPayload data: Data) -> [Entry] {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return [] }
        let window = [root["timeStampBegin"], root["timeStampEnd"]].compactMap { clean($0 as? String, limit: 40) }
            .joined(separator: " – ")
        let crashes = (root["crashDiagnostics"] as? [[String: Any]] ?? []).map { entry(kind: "Crash", diagnostic: $0, window: window) }
        let hangs = (root["hangDiagnostics"] as? [[String: Any]] ?? []).map { entry(kind: "Hang", diagnostic: $0, window: window) }
        return crashes + hangs
    }

    private static func entry(kind: String, diagnostic: [String: Any], window: String) -> Entry {
        let meta = diagnostic["diagnosticMetaData"] as? [String: Any] ?? [:]
        var lines: [String] = []
        var title = kind
        if kind == "Hang", let duration = clean(meta["hangDuration"] as? String, limit: 40) { title += " for \(duration)" }
        if !window.isEmpty { title += " · reported \(window)" }
        lines.append(title)
        let app = clean(meta["appVersion"] as? String, limit: 40).map { version in
            "App \(version)" + (clean(meta["appBuildVersion"] as? String, limit: 40).map { " (\($0))" } ?? "")
        }
        lines.append([app, clean(meta["osVersion"] as? String, limit: 80), clean(meta["platformArchitecture"] as? String, limit: 20)]
            .compactMap { $0 }.joined(separator: " · "))
        if kind == "Crash" {
            var exception: [String] = []
            if let type = integer(meta["exceptionType"]) {
                exception.append("\(machExceptionName(type)) (type \(type), code \(integer(meta["exceptionCode"]).map(String.init) ?? "?"))")
            }
            if let signal = integer(meta["signal"]) { exception.append("signal \(signal)\(signalName(signal).map { " (\($0))" } ?? "")") }
            if !exception.isEmpty { lines.append("Exception: " + exception.joined(separator: " · ")) }
            if let reason = clean(meta["terminationReason"] as? String, limit: 200) { lines.append("Termination: \(reason)") }
            if let region = clean(meta["virtualMemoryRegionInfo"] as? String, limit: 200) { lines.append("Memory: \(region)") }
            if let reason = meta["exceptionReason"] as? [String: Any] {
                let name = clean((reason["exceptionName"] ?? reason["className"]) as? String, limit: 80)
                let message = clean(reason["composedMessage"] as? String, limit: 300)
                let text = [name, message].compactMap { $0 }.joined(separator: ": ")
                if !text.isEmpty { lines.append("Reason: \(text)") }
            }
        }
        let frames = threadFrames(diagnostic["callStackTree"] as? [String: Any])
        if !frames.isEmpty {
            lines.append(kind == "Hang" ? "Main thread:" : "Crashed thread:")
            lines += frames.prefix(maxFrames).enumerated().map { "  \($0.offset) \($0.element)" }
            if frames.count > maxFrames { lines.append("  … \(frames.count - maxFrames) more frames") }
        }
        return Entry(kind: kind, text: String(lines.joined(separator: "\n").prefix(maxEntryCharacters)))
    }

    /// The attributed thread's stack: MetricKit nests each caller in the frame above's `subFrames`.
    private static func threadFrames(_ tree: [String: Any]?) -> [String] {
        let stacks = tree?["callStacks"] as? [[String: Any]] ?? []
        guard let stack = stacks.first(where: { $0["threadAttributed"] as? Bool == true }) ?? stacks.first,
              var frame = (stack["callStackRootFrames"] as? [[String: Any]])?.first else { return [] }
        var frames: [String] = []
        while frames.count <= maxFrames + 64 {
            let binary = clean(frame["binaryName"] as? String, limit: 80) ?? "?"
            let offset = integer(frame["offsetIntoBinaryTextSegment"]).map { "+0x" + String($0, radix: 16) } ?? ""
            let uuid = clean(frame["binaryUUID"] as? String, limit: 36).map { " \($0)" } ?? ""
            frames.append("\(binary) \(offset)\(uuid)".trimmingCharacters(in: .whitespaces))
            guard let next = (frame["subFrames"] as? [[String: Any]])?.first else { break }
            frame = next
        }
        return frames
    }

    private static func clean(_ value: String?, limit: Int) -> String? {
        guard let value else { return nil }
        let text = String(value.unicodeScalars.filter { !CharacterSet.controlCharacters.contains($0) })
            .trimmingCharacters(in: .whitespaces)
        return text.isEmpty ? nil : String(text.prefix(limit))
    }

    private static func integer(_ value: Any?) -> Int64? {
        // NSNumber bridges 0 and 1 to Bool, so test for a JSON boolean by its CF type instead.
        if let number = value as? NSNumber, CFGetTypeID(number) != CFBooleanGetTypeID() { return number.int64Value }
        if let text = value as? String { return Int64(text) }
        return nil
    }

    private static func machExceptionName(_ type: Int64) -> String {
        let names: [Int64: String] = [1: "EXC_BAD_ACCESS", 2: "EXC_BAD_INSTRUCTION", 3: "EXC_ARITHMETIC", 4: "EXC_EMULATION",
                                      5: "EXC_SOFTWARE", 6: "EXC_BREAKPOINT", 7: "EXC_SYSCALL", 8: "EXC_MACH_SYSCALL",
                                      9: "EXC_RPC_ALERT", 10: "EXC_CRASH", 11: "EXC_RESOURCE", 12: "EXC_GUARD", 13: "EXC_CORPSE_NOTIFY"]
        return names[type] ?? "Mach exception"
    }

    private static func signalName(_ signal: Int64) -> String? {
        [4: "SIGILL", 5: "SIGTRAP", 6: "SIGABRT", 7: "SIGEMT", 8: "SIGFPE", 9: "SIGKILL", 10: "SIGBUS",
         11: "SIGSEGV", 12: "SIGSYS", 13: "SIGPIPE", 15: "SIGTERM"][signal]
    }

    /// New summaries join the saved ones; a repeated delivery is kept once, newest last.
    static func merged(_ existing: [Entry], _ new: [Entry]) -> [Entry] {
        var result = existing
        for entry in new where !result.contains(entry) { result.append(entry) }
        return Array(result.suffix(limit))
    }

    /// The export text, or nil when this phone has no crash or hang report.
    static func section(_ entries: [Entry]) -> String? {
        guard !entries.isEmpty else { return nil }
        return (["MagicMobile crash and hang reports (\(entries.count), newest last, from MetricKit on this phone)"]
            + entries.map(\.text)).joined(separator: "\n\n")
    }

    static func load(directory: URL) -> [Entry] {
        let url = directory.appendingPathComponent(fileName)
        guard let size = try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize,
              size <= limit * maxEntryCharacters * 4 + 4_096,
              let data = try? Data(contentsOf: url),
              let entries = try? JSONDecoder().decode([Entry].self, from: data) else { return [] }
        return Array(entries.suffix(limit))
    }

    static func save(_ entries: [Entry], directory: URL) throws {
        try OnDeviceDiagnostics.prepare(directory)
        let data = try JSONEncoder().encode(Array(entries.suffix(limit)))
        let url = directory.appendingPathComponent(fileName)
        #if os(iOS)
        try data.write(to: url, options: [.atomic, .completeFileProtection])
        #else
        try data.write(to: url, options: .atomic)
        #endif
    }

    /// Adds the summaries in `payloads` (MetricKit JSON) and tells open report sheets.
    @discardableResult
    static func record(payloads: [Data], directory: URL) -> Bool {
        let new = payloads.flatMap(entries(fromPayload:))
        guard !new.isEmpty else { return false }
        let existing = load(directory: directory)
        let combined = merged(existing, new)
        guard combined != existing, (try? save(combined, directory: directory)) != nil else { return false }
        NotificationCenter.default.post(name: didChange, object: directory)
        return true
    }

    static func clear(directory: URL) throws {
        let url = directory.appendingPathComponent(fileName)
        if FileManager.default.fileExists(atPath: url.path) { try FileManager.default.removeItem(at: url) }
    }
}

#if canImport(MetricKit) && os(iOS)
import MetricKit

/// Registered at launch (MagicMobileAppDelegate). MetricKit hands the app its own crash and hang
/// diagnostics, usually on the next launch; they are summarized into the local report and never uploaded.
final class OnDeviceCrashReporter: NSObject, MXMetricManagerSubscriber, @unchecked Sendable {
    static let shared = OnDeviceCrashReporter()
    private var started = false

    @MainActor
    func start() {
        guard !started else { return }
        started = true
        MXMetricManager.shared.add(self)
        record(MXMetricManager.shared.pastDiagnosticPayloads)
    }

    func didReceive(_ payloads: [MXDiagnosticPayload]) { record(payloads) }

    private func record(_ payloads: [MXDiagnosticPayload]) {
        let data = payloads.map { $0.jsonRepresentation() }
        guard !data.isEmpty else { return }
        DispatchQueue.main.async {
            OnDeviceSystemReports.record(payloads: data, directory: OnDeviceDiagnostics.defaultDirectory)
        }
    }
}
#endif

File: apps/android/app/src/main/java/io/magicmobile/android/ondevice/OnDeviceDiagnostics.kt
package io.magicmobile.android.ondevice

import android.app.ActivityManager
import android.app.ApplicationExitInfo
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.annotation.RequiresApi
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import io.magicmobile.android.BuildConfig
import io.magicmobile.android.board.BoardSheet
import io.magicmobile.android.board.ConfirmationAction
import io.magicmobile.android.board.ConfirmationDialog
import io.magicmobile.android.ui.BrandTheme
import io.magicmobile.android.ui.IosTextButton
import io.magicmobile.android.ui.MagicPalette
import io.magicmobile.android.ui.SfDesign
import io.magicmobile.android.ui.SfText
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import java.io.File
import java.io.InputStream
import java.time.Instant

/**
 * Port of OnDeviceDiagnostics.swift: the latest local engine failure only, plus this app's own
 * crash, ANR and native-crash summaries from Android's exit history. No snapshots, action payloads,
 * networking, or automatic sharing. Kept in no-backup storage. Main thread only.
 */
class OnDeviceDiagnostics(private val directory: File,
                          private val appVersion: String = "${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})",
                          private val platform: String = OnDeviceSystemReports.platform()) {
    /** What the report sheet shows and shares: the engine failure, then any crash and ANR summaries. */
    var report by mutableStateOf<String?>(null); private set
    var isHistorical by mutableStateOf(false); private set
    var errorMessage by mutableStateOf<String?>(null)
    private var engineReport: String? = null
    private var persistedReport: String? = null
    private var systemReport: String? = null
    private var captureGeneration = 0
    private var clearing = false
    private val file: File get() = File(directory, "latest-engine-failure.txt")

    init {
        try {
            if (file.exists()) {
                if (file.length() > MAX_BYTES) throw IllegalStateException("The saved report is too large.")
                engineReport = file.readText(Charsets.UTF_8)
                persistedReport = engineReport
                isHistorical = true
            }
        } catch (error: Exception) { errorMessage = "Could not load the saved engine report: ${error.message}" }
        systemReport = OnDeviceSystemReports.section(OnDeviceSystemReports.load(directory))
        publish()
    }

    private fun publish() {
        val parts = listOfNotNull(engineReport, systemReport)
        report = if (parts.isEmpty()) null else parts.joinToString("\n\n")
        // Crash and ANR summaries always describe an earlier session.
        if (engineReport == null) isHistorical = systemReport != null
    }

    /** Exit reports recorded after this store opened (the startup reader runs in the background). */
    fun reloadSystemReports() {
        systemReport = OnDeviceSystemReports.section(OnDeviceSystemReports.load(directory))
        publish()
    }

    fun beginAttempt() {
        captureGeneration += 1
        if (report != null) isHistorical = true
    }

    fun save(engineReport: String, status: String) {
        // Reopening the sheet must not date an existing incident as a new failure.
        val current = this.engineReport
        if (current != null && persistedReport == current) {
            val split = current.indexOf("\n\n")
            if (split >= 0 && engineReport.startsWith(current.substring(split + 2))) return
        }
        val text = """
            |MagicMobile local engine diagnostic
            |App: $appVersion
            |OS: $platform
            |Captured: ${Instant.now()}
            |Status: ${status.take(200)}
            |Private: error text may contain card information. Share only by explicit choice.
            |
            |""".trimMargin() + engineReport
        val bytes = boundedUtf8(text, MAX_BYTES)
        this.engineReport = String(bytes, Charsets.UTF_8)
        isHistorical = false
        errorMessage = null
        publish()
        try {
            prepare(directory)
            val temporary = File(directory, "latest-engine-failure.txt.tmp")
            temporary.writeBytes(bytes)
            if (!temporary.renameTo(file)) throw IllegalStateException("Could not replace the saved report.")
            persistedReport = this.engineReport
        } catch (error: Exception) {
            errorMessage = "Report is available to share now, but could not be saved on this device: ${error.message}"
            throw error
        }
    }

    /** Deletes the engine report and the crash and ANR summaries. */
    fun clear() {
        captureGeneration += 1
        if (file.exists() && !file.delete()) throw IllegalStateException("Could not delete the saved report.")
        engineReport = null; persistedReport = null
        OnDeviceSystemReports.clear(directory)
        systemReport = null
        errorMessage = null; isHistorical = false
        publish()
    }

    suspend fun capture(status: String, read: suspend () -> String?) {
        if (clearing) return
        val generation = captureGeneration
        try {
            val report = read() ?: return
            if (generation != captureGeneration) return
            save(report, status)
        } catch (error: Exception) {
            if (generation != captureGeneration) return
            if (errorMessage == null) errorMessage = "Could not capture the local engine report: ${error.message}"
        }
    }

    suspend fun clear(engine: suspend () -> Unit) {
        if (clearing) return
        captureGeneration += 1; clearing = true
        try { engine(); clear() } finally { clearing = false }
    }

    companion object {
        const val MAX_BYTES = 65_536
        private var shared: OnDeviceDiagnostics? = null

        fun directory(context: Context): File = File(context.noBackupFilesDir, "Diagnostics")

        /** The app's one store, like iOS's root @StateObject. Main thread only. */
        fun shared(context: Context): OnDeviceDiagnostics =
            shared ?: OnDeviceDiagnostics(directory(context.applicationContext)).also { shared = it }

        internal fun current(): OnDeviceDiagnostics? = shared

        /** Private to the app; the directory is also outside Android backups. */
        fun prepare(directory: File) {
            if (!directory.isDirectory && !directory.mkdirs()) throw IllegalStateException("Could not create the diagnostics folder.")
        }

        /** UTF-8 bytes of [text], cut at a code point boundary to at most [limit] bytes. */
        fun boundedUtf8(text: String, limit: Int): ByteArray {
            val bytes = text.toByteArray(Charsets.UTF_8)
            if (bytes.size <= limit) return bytes
            var end = limit
            // Continuation bytes are 10xxxxxx: never end inside a multi-byte sequence.
            while (end > 0 && (bytes[end].toInt() and 0xC0) == 0x80) end -= 1
            return bytes.copyOf(end)
        }
    }
}

/**
 * This app's crashes, ANRs and native crashes as Android records them (ApplicationExitInfo,
 * Android 11+), read once at startup and kept only in the diagnostics folder. A summary holds the
 * time, app and Android versions, the system's short description, memory use, the signal of a
 * native crash, and the main thread's frames from an ANR trace. Package, process and device
 * details are left out; the iOS counterpart summarizes MetricKit crash and hang reports.
 */
object OnDeviceSystemReports {
    /** One recorded process exit, copied from ApplicationExitInfo so it can be formatted anywhere. */
    data class Exit(val timestampMillis: Long, val pid: Int, val reason: Int, val status: Int, val description: String?,
                    val pssKb: Long, val rssKb: Long, val trace: String?)

    data class Entry(val kind: String, val id: String, val at: Long, val text: String)

    /** ApplicationExitInfo.REASON_CRASH, REASON_CRASH_NATIVE and REASON_ANR. */
    const val REASON_CRASH = 4
    const val REASON_CRASH_NATIVE = 5
    const val REASON_ANR = 6
    const val FILE_NAME = "crash-and-anr-reports.json"
    const val LIMIT = 5
    const val MAX_ENTRY_CHARACTERS = 8_000
    const val MAX_FRAMES = 32
    private const val MAX_TRACE_BYTES = 512 * 1024

    fun platform(): String = "Android ${Build.VERSION.RELEASE ?: "unknown"} (API ${Build.VERSION.SDK_INT})"

    /** A summary of a crash, ANR or native crash; null for every other exit reason. */
    fun entry(exit: Exit, appVersion: String, platform: String): Entry? {
        val kind = when (exit.reason) {
            REASON_CRASH -> "Crash"; REASON_CRASH_NATIVE -> "Native crash"; REASON_ANR -> "ANR"
            else -> return null
        }
        val lines = ArrayList<String>()
        lines += "$kind · ${Instant.ofEpochMilli(exit.timestampMillis)}"
        lines += "Seen by app $appVersion · $platform"
        if (exit.reason == REASON_CRASH_NATIVE) lines += "Signal ${exit.status}${signalName(exit.status)?.let { " ($it)" } ?: ""}"
        clean(exit.description, 300)?.let { lines += "Reason: $it" }
        if (exit.pssKb > 0 || exit.rssKb > 0) lines += "Memory: PSS ${exit.pssKb / 1024} MB · RSS ${exit.rssKb / 1024} MB"
        exit.trace?.let(::mainThread)?.let { (state, frames) ->
            if (frames.isNotEmpty()) {
                lines += "Main thread${state?.let { " ($it)" } ?: ""}:"
                frames.take(MAX_FRAMES).forEachIndexed { index, frame -> lines += "  $index $frame" }
                if (frames.size > MAX_FRAMES) lines += "  … ${frames.size - MAX_FRAMES} more frames"
            }
        }
        return Entry(kind, "${exit.reason}-${exit.timestampMillis}-${exit.pid}", exit.timestampMillis,
            lines.joinToString("\n").take(MAX_ENTRY_CHARACTERS))
    }

    /** The "main" thread's state and frames from an ANR trace (the text of `traces.txt`). */
    fun mainThread(trace: String): Pair<String?, List<String>>? {
        val lines = trace.lineSequence().iterator()
        while (lines.hasNext()) {
            val header = lines.next()
            if (!header.startsWith("\"main\"")) continue
            val state = header.trim().split(Regex("\\s+")).lastOrNull()?.takeIf { it.isNotEmpty() && !it.contains('=') }
            val frames = ArrayList<String>()
            while (lines.hasNext()) {
                val line = lines.next()
                if (line.isBlank()) break
                val frame = line.trim()
                if (frame.startsWith("at ") || frame.startsWith("- ") || frame.startsWith("native: ")) {
                    clean(frame, 200)?.let(frames::add)
                    if (frames.size > MAX_FRAMES + 64) break
                }
            }
            return state to frames
        }
        return null
    }

    private fun clean(value: String?, limit: Int): String? {
        val text = value?.filter { !Character.isISOControl(it) }?.trim().orEmpty()
        return text.takeIf { it.isNotEmpty() }?.take(limit)
    }

    private fun signalName(signal: Int): String? = mapOf(4 to "SIGILL", 5 to "SIGTRAP", 6 to "SIGABRT", 7 to "SIGBUS", 8 to "SIGFPE",
        9 to "SIGKILL", 11 to "SIGSEGV", 13 to "SIGPIPE", 15 to "SIGTERM")[signal]

    /** New summaries join the saved ones; a repeated exit is kept once; the newest [LIMIT], oldest first. */
    fun merged(existing: List<Entry>, new: List<Entry>): List<Entry> {
        val byID = LinkedHashMap<String, Entry>()
        (existing + new).forEach { byID.putIfAbsent(it.id, it) }
        return byID.values.sortedBy { it.at }.takeLast(LIMIT)
    }

    /** The export text, or null when this phone has no crash or ANR report. */
    fun section(entries: List<Entry>): String? {
        if (entries.isEmpty()) return null
        return (listOf("MagicMobile crash and ANR reports (${entries.size}, newest last, from Android on this phone)") +
            entries.map { it.text }).joinToString("\n\n")
    }

    fun load(directory: File): List<Entry> = runCatching {
        val file = File(directory, FILE_NAME)
        if (!file.exists() || file.length() > LIMIT * MAX_ENTRY_CHARACTERS * 4L + 4_096) return emptyList()
        Json.parseToJsonElement(file.readText(Charsets.UTF_8)).jsonArray.mapNotNull { row ->
            val fields = row.jsonObject
            val kind = fields["kind"]?.jsonPrimitive?.content ?: return@mapNotNull null
            val id = fields["id"]?.jsonPrimitive?.content ?: return@mapNotNull null
            val at = fields["at"]?.jsonPrimitive?.longOrNull ?: return@mapNotNull null
            val text = fields["text"]?.jsonPrimitive?.content ?: return@mapNotNull null
            Entry(kind, id, at, text.take(MAX_ENTRY_CHARACTERS))
        }.takeLast(LIMIT)
    }.getOrDefault(emptyList())

    fun save(entries: List<Entry>, directory: File) {
        OnDeviceDiagnostics.prepare(directory)
        val json = JsonArray(entries.takeLast(LIMIT).map {
            JsonObject(mapOf("kind" to JsonPrimitive(it.kind), "id" to JsonPrimitive(it.id), "at" to JsonPrimitive(it.at), "text" to JsonPrimitive(it.text)))
        })
        val temporary = File(directory, "$FILE_NAME.tmp")
        temporary.writeText(json.toString(), Charsets.UTF_8)
        if (!temporary.renameTo(File(directory, FILE_NAME))) throw IllegalStateException("Could not save the crash reports.")
    }

    /** Adds the summaries of [exits]; true when something new was saved. */
    fun record(exits: List<Exit>, directory: File, appVersion: String, platform: String): Boolean {
        val new = exits.mapNotNull { entry(it, appVersion, platform) }
        if (new.isEmpty()) return false
        val existing = load(directory)
        val combined = merged(existing, new)
        if (combined == existing) return false
        return runCatching { save(combined, directory) }.isSuccess
    }

    fun clear(directory: File) {
        val file = File(directory, FILE_NAME)
        if (file.exists() && !file.delete()) throw IllegalStateException("Could not delete the crash reports.")
    }

    /**
     * Called once from MainActivity: reads this app's recent exits on a background thread. Needs no
     * permission (the app reads only its own history) and sends nothing anywhere.
     */
    fun recordAtStartup(context: Context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) return
        val app = context.applicationContext
        Thread({ runCatching { recordExits(app) } }, "MagicMobile-exit-reports").start()
    }

    @RequiresApi(Build.VERSION_CODES.R)
    private fun recordExits(app: Context) {
        val manager = app.getSystemService(ActivityManager::class.java) ?: return
        val exits = manager.getHistoricalProcessExitReasons(app.packageName, 0, 16)
            .filter { it.reason == REASON_CRASH || it.reason == REASON_CRASH_NATIVE || it.reason == REASON_ANR }
            .map(::exit)
        val appVersion = "${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})"
        if (record(exits, OnDeviceDiagnostics.directory(app), appVersion, platform())) {
            Handler(Looper.getMainLooper()).post { OnDeviceDiagnostics.current()?.reloadSystemReports() }
        }
    }

    @RequiresApi(Build.VERSION_CODES.R)
    private fun exit(info: ApplicationExitInfo): Exit {
        // Only an ANR's trace is text; a native crash's tombstone is binary and left out.
        val trace = if (info.reason == REASON_ANR) runCatching { info.traceInputStream?.use(::readBounded) }.getOrNull() else null
        return Exit(info.timestamp, info.pid, info.reason, info.status, info.description, info.pss, info.rss, trace)
    }

    private fun readBounded(stream: InputStream): String {
        val buffer = ByteArray(MAX_TRACE_BYTES)
        var length = 0
        while (length < buffer.size) {
            val read = stream.read(buffer, length, buffer.size - length)
            if (read < 0) break
            length += read
        }
        return String(buffer, 0, length, Charsets.UTF_8)
    }
}

/**
 * The setup screen's entry to the local report (iOS: "Engine error report" and its sheet). The
 * report appears once there is one; the player reviews it and shares it only by choice.
 */
@Composable
fun OnDeviceDiagnosticsEntry(setup: OnDeviceSetupModel) {
    val context = LocalContext.current
    val store = remember { OnDeviceDiagnostics.shared(context) }
    val scope = rememberCoroutineScope()
    var open by remember { mutableStateOf(false) }
    var confirmDelete by remember { mutableStateOf(false) }
    val failure = setup.errorMessage ?: setup.session.errorMessage
    LaunchedEffect(failure) { if (failure != null) store.capture(setup.status) { setup.diagnosticReport() } }
    if (store.report != null) {
        IosTextButton("Engine error report", { store.beginAttempt(); open = true },
            Modifier.semantics { contentDescription = "Engine error report" }, color = BrandTheme.ember)
    }
    if (open) {
        LaunchedEffect(Unit) { store.capture(setup.status) { setup.diagnosticReport() } }
        BoardSheet({ open = false }, skipPartiallyExpanded = true) {
            Column(Modifier.fillMaxWidth().heightIn(max = 640.dp).verticalScroll(rememberScrollState()).padding(20.dp),
                verticalArrangement = Arrangement.spacedBy(14.dp)) {
                Text("Engine error report", color = MagicPalette.parchment, style = SfText.headline())
                Text("Only the latest engine error report and this app's recent crash and ANR summaries are kept on this phone, " +
                    "excluded from backups. Error text may contain private card information. Nothing is uploaded automatically; " +
                    "review it before sharing.", color = MagicPalette.parchment, style = SfText.footnote())
                store.errorMessage?.let { Text(it, color = Color(1f, 0.45f, 0.4f), style = SfText.footnote()) }
                val report = store.report
                if (report != null) {
                    Text(if (store.isHistorical) "Saved report from an earlier session" else "Latest local incident",
                        color = MagicPalette.parchment, style = SfText.headline())
                    IosTextButton("Share report", {
                        val send = Intent(Intent.ACTION_SEND).apply { type = "text/plain"; putExtra(Intent.EXTRA_TEXT, report) }
                        runCatching { context.startActivity(Intent.createChooser(send, "Share report")) }
                    }, color = BrandTheme.ember, bold = true)
                }
                if (report != null || store.errorMessage != null) {
                    IosTextButton("Delete saved report", { confirmDelete = true }, color = Color(1f, 0.27f, 0.23f))
                }
                if (report != null) Text(report, color = MagicPalette.parchment, style = SfText.caption(design = SfDesign.MONOSPACED))
                else if (store.errorMessage == null) Text("No engine exception has been captured yet. Try starting the match again, then return here if it stops.",
                    color = MagicPalette.parchment, style = SfText.footnote())
                IosTextButton("Done", { open = false }, color = BrandTheme.ember, bold = true)
            }
        }
    }
    if (confirmDelete) {
        ConfirmationDialog("Delete the local engine report?", "This removes the saved report and its in-memory copy. Reports you already shared are not removed.",
            listOf(ConfirmationAction("Delete report", destructive = true) {
                scope.launch {
                    try { store.clear { setup.clearDiagnostics() } }
                    catch (error: Exception) { store.errorMessage = "Could not delete the local engine report: ${error.message}" }
                }
            })) { confirmDelete = false }
    }
}
